source/cellular_bg770_api.c

```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
+            if( pFrequencyBands->nbIotBands_hexString[ sizeof ( pFrequencyBands->nbIotBands_hexString ) - 1 ] != '\0' )
+            {
+                pFrequencyBands->nbIotBands_hexString[ sizeof ( pFrequencyBands->nbIotBands_hexString ) - 1 ] = '\0';
+                LogError( ( "_parseFrequencyBands: nbIotBands string truncation. Token '%s' nbIotBands_hexString '%s'",
+                        pToken, pFrequencyBands->nbIotBands_hexString ) );
+                parseStatus = false;
+            }
+        }
+        else
+        {
+            LogError( ( "_parseFrequencyBands: nbIotBands string not present" ) );
+            pFrequencyBands->nbIotBands_hexString[ 0 ] = '\0';
+            parseStatus = false;
+        }
+    }
+
+    return parseStatus;
+}
+
+static void _removeHexValuePrefixIfPresent(const char **pString) {
+    static const char *const HEX_PREFIX_1 = "0x";
+    static const char *const HEX_PREFIX_2 = "0X";
+    static const size_t HEX_PREFIX_LENGTH = 2;
+
+    if( pString == NULL || *pString == NULL )
+    {
+        LogError( ( "NULL string in _removeHexValuePrefixIfPresent" ) );
+        return;
+    }
+
+    if( strncmp( *pString, HEX_PREFIX_1, HEX_PREFIX_LENGTH ) == 0 ||
+        strncmp( *pString, HEX_PREFIX_2, HEX_PREFIX_LENGTH ) == 0 )
+    {
+        *pString += HEX_PREFIX_LENGTH;
+    }
+}
+
+// NOTE: can handle "0x" prefix in hex string
+static bool _convertLTEFrequencyBandHexStringToBandBitmask( const char *const pBandHexString,
+                                                            const size_t bandHexStringMaxLength,
+                                                            CellularLTEBandMask_t *const pLTEBandMask )
+{
+    if( ( pBandHexString == NULL ) ||  ( pLTEBandMask == NULL ) || ( bandHexStringMaxLength == 0 ) )
+    {
+        return false;
+    }
+
+    size_t bandHexStringLen = strnlen( pBandHexString, bandHexStringMaxLength );
+    if( bandHexStringLen == bandHexStringMaxLength )
+    {
+        // expects a null terminated string
+        return false;
+    }
+
+    const char * pTempBandHexString = &pBandHexString[ 0 ];
+    _removeHexValuePrefixIfPresent( &pTempBandHexString );
+
+    static const size_t LTE_BAND_MASK_LENGTH = sizeof( pLTEBandMask->asBytes );
+    bandHexStringLen = strnlen( pTempBandHexString, bandHexStringMaxLength - 1 );
+    if( bandHexStringLen > LTE_BAND_MASK_LENGTH * 2u )
+    {
+        // hex string doesn't fit in LTE band mask
+        return false;
+    }
+
+    memset( pLTEBandMask->asBytes, 0x00, LTE_BAND_MASK_LENGTH );
+
+    // loop through nibbles and apply to band bitmask
+    unsigned int byteIndex = LTE_BAND_MASK_LENGTH - 1;
+    bool isUpperNibble = false;
+    for( int i = bandHexStringLen - 1; i >= 0; i-- )
+    {
+        char tempBuffer[ 2 ] = { pTempBandHexString[ i ], '\0' };
+        char *pEnd = tempBuffer;
+        long tempVal = strtol( tempBuffer, &pEnd, 16 );

```

Comment:
pEnd is useful since it allows differentiation between 0 because of error and real zero. pEnd == NULL or *pEnd not null terminator indicates error conditions since we cannot look at unsafe errno.